server/functions/0_compiled_function.go

```diff
@@ -0,0 +1,264 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	"github.com/dolthub/vitess/go/vt/proto/query"
+	"github.com/shopspring/decimal"
+)
+
+// CompiledFunction is an expression that represents a fully-analyzed PostgreSQL function.
+type CompiledFunction struct {
+	Name       string
+	Parameters []sql.Expression
+	Functions  *OverloadDeduction
+}
+
+var _ sql.FunctionExpression = (*CompiledFunction)(nil)
+
+// FunctionName implements the interface sql.Expression.
+func (c *CompiledFunction) FunctionName() string {
+	return c.Name
+}
+
+// Description implements the interface sql.Expression.
+func (c *CompiledFunction) Description() string {
+	return fmt.Sprintf("The PostgreSQL function `%s`", c.Name)
+}
+
+// Resolved implements the interface sql.Expression.
+func (c *CompiledFunction) Resolved() bool {
+	for _, param := range c.Parameters {
+		if !param.Resolved() {
+			return false
+		}
+	}
+	return true
+}
+
+// String implements the interface sql.Expression.
+func (c *CompiledFunction) String() string {
+	sb := strings.Builder{}
+	sb.WriteString(c.Name + "(")
+	for i, param := range c.Parameters {
+		if i > 0 {
+			sb.WriteString(", ")
+		}
+		sb.WriteString(param.String())
+	}
+	sb.WriteString(")")
+	return sb.String()
+}
+
+// OverloadString returns the name of the function represented by the given overload.
+func (c *CompiledFunction) OverloadString(types []IntermediateParameter) string {
+	sb := strings.Builder{}
+	sb.WriteString(c.Name + "(")
+	for i, t := range types {
+		if i > 0 {
+			sb.WriteString(", ")
+		}
+		sb.WriteString(t.CurrentType.String())
+	}
+	sb.WriteString(")")
+	return sb.String()
+}
+
+// Type implements the interface sql.Expression.
+func (c *CompiledFunction) Type() sql.Type {
+	if resolvedFunction, _ := c.Functions.ResolveByType(c.possibleParameterTypes()); resolvedFunction != nil {
+		return resolvedFunction.ReturnSqlType
+	}
+	// We can't resolve to a function before evaluation in this case, so we'll return something arbitrary
+	return types.LongText
+}
+
+// IsNullable implements the interface sql.Expression.
+func (c *CompiledFunction) IsNullable() bool {
+	// We'll always return true, since it does not seem to have a truly negative impact if we return true for a function
+	// that will never return NULL, however there is a negative impact for returning false when a function does return
+	// NULL.
+	return true
+}
+
+// Eval implements the interface sql.Expression.
+func (c *CompiledFunction) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	// First we'll evaluate all of the parameters.
+	parameters := make([]IntermediateParameter, len(c.Parameters))
```

Comment:
Good rule of thumb for anything this complicated: extract each type switch and necessary surrounding code as its own method